File: platforms/swift/Sources/ShopifyCheckoutKit/CheckoutEventAdapter.swift
#if !COCOAPODS
    import EmbeddedCheckoutProtocol
#endif
import Foundation

@MainActor
protocol CheckoutEventSink: AnyObject {
    func checkoutDidStart(_ checkout: Checkout)
    func checkoutDidUpdate(_ checkout: Checkout)
    func checkoutDidComplete(_ checkout: Checkout)
}

/// Translates protocol notifications into the stable lifecycle exposed by Checkout Kit.
struct CheckoutEventAdapter: CheckoutCommunicationProtocol {
    private let base: (any CheckoutCommunicationProtocol)?
    private let eventClient: EmbeddedCheckoutProtocol.Client

    @MainActor
    init(base: (any CheckoutCommunicationProtocol)? = nil, sink: any CheckoutEventSink) {
        self.base = base
        let state = CheckoutEventState(sink: sink)
        eventClient = EmbeddedCheckoutProtocol.Client()
            .on(CheckoutProtocol.start) { state.start($0) }
            .on(CheckoutProtocol.lineItemsChange) { state.update($0) }
            .on(CheckoutProtocol.messagesChange) { state.update($0) }
            .on(CheckoutProtocol.totalsChange) { state.update($0) }
            .on(CheckoutProtocol.fulfillmentChange) { state.update($0) }
            .on(CheckoutProtocol.complete) { state.complete($0) }
    }

    func process(_ message: String) async -> String? {
        let response = await base?.process(message)
        let eventResponse = await eventClient.process(message)
        return response ?? eventResponse
    }
}

@MainActor
private final class CheckoutEventState {
    weak var sink: (any CheckoutEventSink)?
    private var latestCheckout: Checkout?

    init(sink: any CheckoutEventSink) {
        self.sink = sink
    }

    func start(_ checkout: some Encodable) {
        guard let snapshot = Checkout(protocolCheckout: checkout) else { return }
        latestCheckout = snapshot
        sink?.checkoutDidStart(snapshot)
    }

    func update(_ checkout: some Encodable) {
        guard let snapshot = Checkout(protocolCheckout: checkout) else { return }
        guard snapshot != latestCheckout else { return }
        latestCheckout = snapshot
        sink?.checkoutDidUpdate(snapshot)
    }

    func complete(_ checkout: some Encodable) {
        guard let snapshot = Checkout(protocolCheckout: checkout) else { return }
        latestCheckout = snapshot
        sink?.checkoutDidComplete(snapshot)
    }
}

File: platforms/android/lib/src/main/java/com/shopify/checkoutkit/CheckoutEventAdapter.kt
package com.shopify.checkoutkit

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import com.shopify.ucp.embedded.checkout.Checkout as ProtocolCheckout

/** Translates supported protocol notifications for one checkout presentation. */
internal class CheckoutEventAdapter(
    private val listener: CheckoutWebViewListener,
    onDecodeError: (String) -> Unit = {},
) {
    private var active = true
    private var latestSnapshot: JsonElement? = null
    private val client = CheckoutProtocol.Client()
        .on(CheckoutProtocol.start, ::start)
        .on(CheckoutProtocol.lineItemsChange, ::update)
        .on(CheckoutProtocol.messagesChange, ::update)
        .on(CheckoutProtocol.totalsChange, ::update)
        .on(CheckoutProtocol.fulfillmentChange, ::update)
        .on(CheckoutProtocol.complete, ::complete)
        .withDecodeErrorObserver(onDecodeError)

    fun process(message: String) {
        client.process(message)
    }

    /** Called on the main thread before replacing or releasing the presentation listener. */
    fun invalidate() {
        active = false
        latestSnapshot = null
    }

    fun actionForLink(link: CheckoutLink): CheckoutLinkAction =
        if (active) listener.onCheckoutLinkClicked(link) else CheckoutLinkAction.Cancel

    private fun start(checkout: ProtocolCheckout) {
        if (!active) return
        val snapshot = Checkout.fromProtocol(checkout)
        remember(snapshot)
        listener.onCheckoutStarted(CheckoutStartEvent(snapshot))
    }

    private fun update(checkout: ProtocolCheckout) {
        if (!active) return
        val snapshot = Checkout.fromProtocol(checkout)
        val comparison = Json.encodeToJsonElement(Checkout.serializer(), snapshot)
        if (comparison == latestSnapshot) return
        latestSnapshot = comparison
        listener.onCheckoutUpdated(CheckoutUpdateEvent(snapshot))
    }

    private fun complete(checkout: ProtocolCheckout) {
        if (!active) return
        val snapshot = Checkout.fromProtocol(checkout)
        remember(snapshot)
        listener.onCheckoutCompleted(CheckoutCompleteEvent(snapshot))
    }

    private fun remember(checkout: Checkout) {
        // Keep comparison state separate from collections exposed to Kotlin and Java consumers.
        latestSnapshot = Json.encodeToJsonElement(Checkout.serializer(), checkout)
    }
}
